_config/security.yml

```diff
@@ -42,3 +43,11 @@ SilverStripe\Core\Injector\Injector:
       Authenticators:
         cms: '%$SilverStripe\Security\MemberAuthenticator\CMSMemberAuthenticator'
   SilverStripe\Security\IdentityStore: '%$SilverStripe\Security\AuthenticationHandler'
+
+SilverStripe\Security\PasswordExpirationMiddleware:
+  default_redirect: /Security/changepassword
```

Comment:
I think removing the slash from the start of these urls should let you use the absoluteLink() method the way you expect, right?

Reply:
I disagree. This would be semantically incorrect.

According to [RFC1808, Relative Uniform Resource Locators](https://tools.ietf.org/html/rfc1808#section-2.2) lack of leading slash would indicate the redirect should be relative to the request URI, which should not be the case in this particular case.

As such, `Director::absoluteURL` method in its current implementation is not appropriate for the current scenario.

Reply:
I actually read that RFC and it completely backs up what I have been trying to tell you.

![image](https://user-images.githubusercontent.com/936064/64568062-e79ade00-d3ad-11e9-80ee-b2fe14c79f32.png)

>  (3.1) Base URL embedded in the document's content 

This is Director::baseURL().

Please remove the leading slashes and the custom url resolution logic.

Reply:
> (3.1) Base URL embedded in the document's content

Yes, agree, however that's not what I'm talking about.

1. Middlewares work out of context of documents, since routing hasn't been done yet. There is no document identified for serving yet, so `3.3` applies, and not `3.1`.

> (3.3) URL used to retrieve the entity

If we ignore that and say that `SS_BASE_URL` must be applied everywhere and for everything in the framework, then we cannot remove the custom logic, because Director::absoluteURL doesn't follow that rule.

Reply:
> Director::absoluteURL doesn't follow that rule

To elaborate - simply using it would produce incorrect results for absolute paths, which are relative URLs. As long as we don't have any rules about prohibiting use of leading slashes in the config files, I think that would be a wrong decision to make, leading to bugs in users projects.

Reply:
Just had a read through both points here, and I'm leaning toward what @dnsl48 is saying. To summarise:

When given a leading slash, `Director::absoluteURL()` becomes a framework-agnostic utility function for generating an absolute URL against the current request. This completely subverts the `SS_BASE_URL` setting, which is trouble for those who are:

* Using a canonical domain name
* Running their project in a subdirectory

Given that, I don't think 3.1 of the RFC applies here, because no base URL context provided, in the document, nor in the middleware. Further, the two above conditions make the `Director::absoluteURL()` function unreliable in some contexts.

Briefly, if we use `Director::absoluteURL()`, we open up a window for errors and unexpected behaviour, and if we don't, we may have some redundant code, depending on how you interpret an RFC. I prefer the latter.

Reply:
> When given a leading slash, Director::absoluteURL() becomes a framework-agnostic utility function for generating an absolute URL against the current request. This completely subverts the SS_BASE_URL setting, which is trouble for those who are:

This is why I instructed him to remove the leading slashes. Remove the leading slashes, and you have no problems. There's no question about how leading slashes should be handled, they are absolute path according to the RFC. Even @dnsl48 understood this.

The question is about how relative paths should be handled, which is the point we disagree with, hence his apprehension in using them.

>  Further, the two above conditions make the Director::absoluteURL() function unreliable in some contexts.

It's only unreliable to users who don't (or refuse) to understand.

Relative paths could be handled one of two ways (second argument to absoluteURL()):

 - relative to application base path (rfc behavioru in presence of a base url, preferred behaviour in 99% of times this method is called)
 - relative to current requets (rfc behaviour in absence of base url, but virtually never the desired outcome of devs calling this method).

Needless the say, the current 4.x default behaviour is to default in favour of the base path as the relative parent, not the current request. That's not a matter of opinion, it's a fact.

If this is to change, then that's another topic you are welcome to discuss in the context of 5.x, and potentially in the context of removing base_url entirely (something which many devs desire in the long term).

However, in the case of getting a new feature merged into the 4 branch of framework, you need to both understand and accept the current convention for this major feature. Changing an existing url resolution pattern is irrelevant in the context of getting this feature merged.

Reply:
> If we ignore that and say that SS_BASE_URL must be applied everywhere and for everything in the framework, then we cannot remove the custom logic, because Director::absoluteURL doesn't follow that rule.

No, only for relative url resolution. Not "absolutely everything". You are intentionally misrepresenting what I said.

Reply:
> To elaborate - simply using it would produce incorrect results for absolute paths, which are relative URLs. As long as we don't have any rules about prohibiting use of leading slashes in the config files, I think that would be a wrong decision to make, leading to bugs in users projects.

This is why I said not to use absolute paths. You surely read the first of my very easy change requests right?

Reply:
> You are intentionally misrepresenting what I said.

Sorry if I did that, it wasn't intentional. I meant to say about it in a general way as an opposite to the other option and wasn't trying to refer to your words there.

> The question is about how relative paths should be handled, which is the point we disagree with

Sorry, but that's not where I'm disagreeing with you.
I rather consider `Remove the leading slashes, and you have no problems` `It's only unreliable to users who don't (or refuse) to understand` to be a sub-optimal solution.

Configuring the `default redirect path` for the middleware people won't know how exactly it's gonna be resolved. They don't have to know details of Director::absoluteURL implementation and most likely they won't be even knowing that PasswordExpirationMiddleware is using it internally.

Users are going to be putting the leading slash into the configs because it feels semantically right according to the web standards. By removing the custom logic around absolute url resolutions in this middleware we would introduce a surface for bugs occurring there.

I can remove the leading slash from the config, since it's closer to our current conventions (which aren't documented by the way, that's why I'm one of those who don't know about them). However, it feels like (and I've already said that I actually don't really know them) those conventions are rather about routing and not completely applicable to redirects. Routing doesn't really have to deal with how absolute vs relative paths should be handled. Redirects do.

I think Director::absoluteURL is a red herring here. I'd rather be thinking about how the middleware should work and what it should be doing with its configuration.
So far I've got tests covering both use cases - lack of the leading slash and its presence. As long as the conventions are not rules and we don't prohibit leading slashes, I believe it would be more flexible and less error prone to handle both gracefully rather than remove support for absolute paths and demand people not to use leading slashes.